Add determinism developer documentation (#5718)
NVIDIA/Megatron-LM: docs/developer/determinism/status.md

> **Audience:** Megatron-LM / Megatron-Core developers working on deterministic
> training. For setup and operational constraints, see the
> [Deterministic Training user guide](../../user-guide/deterministic-training.md).
>
> **Companion docs in this directory**
> - [`validation-evidence.md`](./validation-evidence.md) — detailed findings,
>   conclusions, and explanations from the validation work.
> - [`training-path.md`](./training-path.md) — the forward+backward branch map:
>   every point in the training step where the code (or an underlying kernel)
>   chooses a deterministic vs non-deterministic path.
> - [`op-catalog.md`](./op-catalog.md) — the per-op catalog table (det? / det path /
>   non-det path / how selected / perf Δ / gap).

---

## 1. What "deterministic" means here

**Bitwise determinism:** given a fixed input and seed, the model yields an
*identical* loss/grad curve across every run, with everything else held fixed —
input data + order + seed, model recipe + config, parallelism layout
(TP/PP/DP/EP/CP/VPP), container image, library versions (MCore, CUDA, cuDNN,
NCCL, Transformer Engine, PyTorch, driver), NCCL env/runtime settings, and

> hardware type + cluster topology.

Reviewer: emm shall we move it to `glossary.md`

  fallback paths than non-deterministic mode — this is the design we use.

## 2. Why it is hard

The primary enemy is **floating-point non-associativity**: `(a+b)+c ≠ a+(b+c)` in
bf16/fp16/fp32. Any kernel whose result depends on the *order* of a parallel
reduction can produce different round-off between runs. The usual culprits are
atomic adds, multi-block reductions, async collectives whose completion order
varies, and allocator-driven kernel autotuning. Determinism therefore means
pinning reduction order everywhere it matters.

## 3. Targets and motivation

Customers are moving determinism from a debug/CI nicety to a **production / hero-run
requirement**. They want to replay loss spikes deterministically, resume from a
checkpoint along the original trajectory, and have a proof point that the GPU
stack produces correct math, which mitigates SDC/SDE concerns. The blocker to
enabling it by default is the performance penalty.

| Milestone | Determinism overhead (step time) |
| --- | --- |
| Current baseline (mcore) | **~15%** overall; Nemotron-3-Ultra **+17.4%**, DSV3 **+8.5%** |
| Adoption target | **< 10%** (would let us default `deterministic_mode=True`) |

> | Aggressive target | **≈ 5%** (per Meituan Longcat: deterministic FAG + ScatterAdd + grouped GEMM + fused GemmAdd) |

Reviewer: can we avoid mentioning `per Meituan Longcat:`?

- Full deterministic overhead remains workload- and allocation-sensitive.
  Attention backward, vocab-parallel loss/gather backward, and grouped-GEMM
  paths can dominate after the routing and reduction improvements land.

The current performance result is therefore a direction, not a global claim.
Each optimization needs an exactness comparison and an alternating-order
end-to-end measurement before it is accepted.

## 4. Control plane — how determinism is turned on

### 4.1 Upstream state

PR #5041 **merged into `NVIDIA/main` on 2026-07-08**, so the determinism test
scaffold (`BitExactRunner`, the det-vs-nondet nsys leaderboard, and the
`determinism.py` argument helpers) is upstream. This branch was stacked on that
PR's pre-merge head and adds the implementation, tests, diagnostics, and
evidence described below; those additional commits have **no separate upstream
PR yet** and are not an upstream guarantee. A rebase onto post-merge `main` is
pending.

The config flag is `model_parallel_config.py` `deterministic_mode: bool = False`,
threaded into `TransformerConfig`; library code reads that flag or
`torch.are_deterministic_algorithms_enabled()`.

### 4.2 This branch

> `--deterministic-mode` delegates to reusable setup helpers:

Reviewer: Can we reference the existing determinism guide instead of duplicating the same information? That would make maintenance easier.

docs/user-guide/deterministic-training.md

Feel free to relocate the above docs to a better place.

- **`megatron.training.determinism.apply_determinism_to_args(args)`** rejects CE
  fusion, replaces uncertified DeepEP/HybridEP dispatch with standard all-to-all,
  validates `NCCL_ALGO` (excluding `Tree`), forces `tp_comm_overlap=False`, and
  finally enables torch deterministic algorithms.
  For the distributed optimizer it also requires one optimizer instance,
  rejects collective AVG, and forces ordered fp32 reduce-scatter.
- **Small floating-point statistics** (gradient norm, reported loss, and MoE
  metrics) use rank-ordered all-gather plus a fixed local sum. `NCCL_ALGO=Ring`
  selects an algorithm but does not pin the physical ring across allocations.

Flash attention is permitted when Transformer Engine honors
`NVTE_ALLOW_NONDETERMINISTIC_ALGO=0`; the bit-exact suite covers supported
configurations.

## 5. Enforced limitations in deterministic mode

These features are currently **incompatible** with deterministic mode:

| Feature | Where enforced | Reason |
| --- | --- | --- |
| `cross_entropy_loss_fusion` | `megatron/training/determinism.py` assert (`apply_determinism_to_args`, reached from `arguments.py` `validate_args`) | Fused CE kernel is non-deterministic |

> | Flex MoE dispatcher (DeepEP/HybridEP) | `determinism.py` switches to standard all-to-all; `TransformerConfig` rejects any surviving flex config | External fused dispatch/combine kernels do not yet have a fixed-order cross-allocation certificate |

Reviewer: > `determinism.py` switches to standard all-to-all

Is there a reference for this?

   `compare_dumps.py` localizes saved activations, parameters, and gradients.
   The structured trace covers phase ordering, recompute correspondence,
   optimizer boundaries, completed collective boundaries, and selected
   attention backends. It does not yet independently capture Transformer
   Engine FP8/FP4 quantizer state, tensor-parallel userbuffer payloads, or
   selection state for arbitrary auto-dispatching libraries.

4. **Some floating reductions still lack topology-independent paths.**
   Tensor-parallel mappings and linears, vocab-parallel cross entropy, and the
   non-distributed-optimizer data-parallel all-reduce remain conditional or
   unsupported at the required scope. Megatron-FSDP, fused cross entropy,
   tensor-parallel communication overlap, multiple distributed optimizers,
   collective averaging, and packed-sequence SSM remain intentionally blocked
   in deterministic mode.

## 9. References

- Determinism roadmap & meeting notes (internal Google Docs).
- Customer technical reports: Meituan Longcat (deterministic FAG / ScatterAdd /
  grouped GEMM / fused GemmAdd), MSFT, DeepSeek-V4.
- Thinking Machines, "Defeating nondeterminism in LLM inference."

> - PyTorch reproducibility & `torch.use_deterministic_algorithms` docs.

Reviewer: emm maybe we don't need to add these.

---
orphan: true
---

> # Determinism status

Reviewer 2: ```suggestion
# Determinism Status
```

> > **Audience:** Megatron developers. Setup and supported configurations are in

Reviewer 2: ```suggestion
> 
```

> > the [user guide](../../user-guide/deterministic-training.md); definitions are

Reviewer 2: ```suggestion
> 
```

## Deterministic Mode

> `--deterministic-mode` (refer to `megatron/training/determinism.py`) validates the

Reviewer 2: ```suggestion
`--deterministic-mode` (refer to `megatron/training/determinism.py`) does the
following:
```

Author: Applied in 257e0ebf2.

> > in the [glossary](./glossary.md); per-operation detail is in the

Reviewer 2: ```suggestion
> 
```

> > [op catalog](./op-catalog.md). The live roadmap is

Reviewer 2: ```suggestion
> 
```

> determinism environment variables and fills canonical defaults, rejects the

Reviewer 2: ```suggestion

- Validates the determinism environment variables and fills canonical defaults
- Rejects features that have no deterministic path (cross-entropy fusion and
  tensor parallelism (TP) communication overlap)
- Enables `torch.use_deterministic_algorithms(True)`
```

Author: Applied in 257e0ebf2.

> > [issue #5785](https://github.com/NVIDIA/Megatron-LM/issues/5785).

Reviewer 2: ```suggestion
>
```

> features that have no deterministic path (cross-entropy fusion and TP

Reviewer 2: ```suggestion
```

Author: Applied in 257e0ebf2.

> ## What deterministic mode does today

Reviewer 2: ```suggestion
## Deterministic Mode
```

> communication overlap), and enables `torch.use_deterministic_algorithms(True)`.

Reviewer 2: ```suggestion
```

> `--deterministic-mode` (see `megatron/training/determinism.py`) validates the

Reviewer 2: ```suggestion
`--deterministic-mode` (refer to `megatron/training/determinism.py`) validates the
```

> The library code then selects the deterministic branches listed in the

Reviewer 2: ```suggestion
```

> [op catalog](./op-catalog.md). Refer to the user guide for exact flags and

Reviewer 2: ```suggestion
```

> environment values.

Reviewer 2: ```suggestion

The library code then selects the deterministic branches listed in the
[Op Catalog](./op-catalog.md). Refer to the User Guide for exact flags and
environment values.
```

## Validation

> 

Reviewer 2: ```suggestion

Deterministic mode uses the following verification layers:

```

> Library code then selects the deterministic branches listed in the

Reviewer 2: ```suggestion
The library code then selects the deterministic branches listed in the
```

> [op catalog](./op-catalog.md). The user guide documents the exact flags and

Reviewer 2: ```suggestion
[op catalog](./op-catalog.md). Refer to the user guide for exact flags and
```

> environment values — they are not repeated here.

Reviewer 2: ```suggestion
environment values.
```

- **Module-level bit-exact suite** (`tests/unit_tests/determinism/`): Runs a
  model or block twice under restored RNG state and asserts bit-identical
  outputs and gradients. Coverage includes:

  - GPTModel, TransformerBlock, and HybridModel

>   - Tensor parallelism, expert parallelism, fully sharded data parallel, pipeline

Reviewer 2: ```suggestion
  - Tensor parallelism, expert parallelism, and fully sharded data parallel
```

> - **Module-level bit-exact suite** (`tests/unit_tests/determinism/`): runs a

Reviewer 2: ```suggestion
- **Module-level bit-exact suite** (`tests/unit_tests/determinism/`): Runs a
  model or block twice under restored RNG state and asserts bit-identical
  outputs and gradients. Coverage includes:

  - GPTModel, TransformerBlock, and HybridModel
  - Tensor parallelism, expert parallelism, fully sharded data parallel, pipeline
    parallelism, and virtual pipeline parallelism
  - FP8 and FP4 recipes
  - Scheduling stressors to surface latent ordering races
- **Performance gate**
  (`tests/performance_tests/shell_test_utils/determinism/`): Runs a small
  recipe in deterministic and default mode under Nsight Systems, reports a
  per-range leaderboard, and fails when the deterministic step time exceeds the
  documented threshold.
- **End-to-end verification**: Compares full-precision training
  metrics across two independent runs (refer to the glossary's "Verification"
  note). Extending checked-in coverage to production-scale architectures is a
  roadmap item.
```

Reviewer 2: delete lines 26-38

>     parallelism, and virtual pipeline parallelism

Reviewer 2: ```suggestion
  - Pipeline parallelism and virtual pipeline parallelism
```

  - FP8 and FP4 recipes
  - Scheduling stressors to surface latent ordering races
- **Performance gate**
  (`tests/performance_tests/shell_test_utils/determinism/`): Runs a small
  recipe in deterministic and default mode under Nsight Systems, reports a

>   per-range leaderboard, and fails when the deterministic step time exceeds the

Reviewer 2: ```suggestion
  leaderboard for each range, and fails when the deterministic step time exceeds the
```

  documented threshold.
- **End-to-end verification**: Compares full-precision training metrics across
  two independent runs (refer to the glossary's "Verification" note). Extending
  checked-in coverage to production-scale architectures is a roadmap item.

## Performance

Deterministic mode increases step time by roughly 15%, varying by model and

> precision. The goal is under 10%, with a stretch goal near 5%, so you can leave

Reviewer 2: ```suggestion
precision. The goal is under 10%, with a stretch goal near five percent, so you can leave
```

> determinism on in production runs. The hotspot list and optimization progress

Reviewer 2: ```suggestion
determinism on in production runs. 
```

> Deterministic mode currently costs roughly 15% step time, varying by model and

Reviewer 2: ```suggestion
Deterministic mode increases step time by roughly 15%, varying by model and
precision. The goal is under 10%, with a stretch goal near five percent, so you can leave determinism on in production runs. The hotspot list and optimization progress
live in the [Operations Catalog](./op-catalog.md) and
[issue #5785](https://github.com/NVIDIA/Megatron-LM/issues/5785).
```

Reviewer 2: and delete 43-46

> live in the [op catalog](./op-catalog.md) and

Reviewer 2: ```suggestion

Refer to the [Op Catalog](./op-catalog.md) and
[issue #5785](https://github.com/NVIDIA/Megatron-LM/issues/5785) for the hotspot
list and optimization progress.
```

> [issue #5785](https://github.com/NVIDIA/Megatron-LM/issues/5785).

Reviewer 2: ```suggestion
```
